src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIContainerService.java

```
@@ -402,17 +398,34 @@ public Path createBuildScript(ProgrammingExerciseParticipation participation, St
         buildScript.append("#!/bin/bash\n");
         buildScript.append("cd ").append(WORKING_DIRECTORY).append("/testing-dir\n");
 
-        actions.forEach(action -> {
-            String workdir = action.getWorkdir();
-            if (workdir != null) {
-                buildScript.append("cd ").append(WORKING_DIRECTORY).append("/testing-dir/").append(workdir).append("\n");
+        String customScript = programmingExercise.getBuildScript();
+        if (customScript != null) {
+            buildScript.append(customScript);
+        }
+        else {
+            List<ScriptAction> actions = List.of();
+
+            Windfile windfile = programmingExercise.getWindfile();
+
+            if (windfile == null) {
+                windfile = aeolusTemplateService.getDefaultWindfileFor(programmingExercise);
             }
-            buildScript.append(action.getScript()).append("\n");
-            if (workdir != null) {
-                buildScript.append("cd ").append(WORKING_DIRECTORY).append("/testing-dir\n");
+            if (windfile != null) {
```

Maybe throw an exception if for any reason windfile is still null?